### DISCUSSION TITLE
Way to allow for two different comment out methods for Lua scripts

### DISCUSSION THREAD

**Discussion by advice2020** [geany/geany#3997] [Q&A]
In Lua scripts the common symbol used to comment items out single lines is `--`
However you can also use `--[[` to begin and `]]` to close for commenting out multiple lines (similar to XML comment)
I was wondering if there was any way to achieve anything like this in Geany?

Where if cursor is in a line and you use comment out hotkey it adds the `--` but if multiple lines are highlighted and hotkey is applied that it uses the brackets method to comment them out?

Or even a separate hotkey one for each

I have seen that you can modify the comment out symbols in `filetypes.xml.conf` but I think there is only one line for open and one line for close, so this is probably not possible to split this.

If anyone has any ideas please let me know.
Thank You

**Comment by elextr** [comment]
Lua comments are defined in the filetype as `--[[` and `--]]`.

Commands that can have keybindings can be discovered in the `Edit->Preferences->Keybindings`, saying "hotkey" is meaningless, post the command you actually used.

IIRC the the `Comment line(s)` menu uses single line comments if defined in the filetype else multi line comments if defined.  You can check the Geany-plugins collection if that has more capabilities, or you could always create your own plugin.

**Comment by advice2020** [comment]
Sorry formatting issue.
`--[[` and `]]` work for me but yours would as well because the `--` at the end would be inside of the brackets.

Not sure saying "hotkey" is "meaningless", aren't they all linked to only ONE commenting action on/off?
Anyways, I use the default hotkey which toggles commenting on/off.
`Toggle Line Commentation = Ctrl+E`

I found the `filetypes.lua` file which shows
```
# single comments, like # in this file
#comment_single=--
# multiline comments
#comment_open=--[[
#comment_close=--]]
```
So it is set for single / multiple lines but I do not know how the multiline comment is activated?
If cursor is in the line and I hit Ctrl+E it add `--` to the beginning
and if I select multiple lines and apply Ctrl+E it just adds `--` to the beginning of each line.

Are you were saying that if a "filetypes" file does contain a "#comment_single=" value, then this then overrides the "multiline" value?  Does appear to be true because I changed the following line (removed # and --)
#comment_single=--
to
comment_single=
and then it only uses the --[[ and --]] commenting.

Does anyone know if it is possible to use both single and multiline or is it only one or the other?

Thank You

**Comment by elextr** [comment]
> Not sure saying "hotkey" is "meaningless", aren't they all linked to only ONE commenting action on/off?

"Hotkey" does not tell me what command you did.

> I use the default hotkey which toggles commenting on/off. `Toggle Line Commentation = Ctrl+E`

And now readers actually know what command you used.

> Does anyone know if it is possible to use both single and multiline or is it only one or the other?

I havn't time to examine the details of Geany, but as I said AFAIK, its single line if a sensible value is present, else multiple line.

As I said you should check if any of the plugins provide more unless somebody else knows.

**Comment by advice2020** [comment]
Thank you for your response.
I thought all the "comment" based hotkeys all performed the same action of comment/uncomment which as mentioned was why I just said "hotkey" implying the hotkey used to activate the "comment" action, sorry about that.

I apologize I do not know what you mean when you say "sensible value is present"
You had originally said "menu uses single line comments if defined in the filetype else multi line comments"
I viewed this as what I mentioned that I thought you meant by this
If the "single line" line in the filetype file did not have defined "symbols" in it meaning if left blank then it would use the "symbols" shown in the "multi line" line in the filetype file.
As mentioned this seemed to be true in my testing but was just asking to see if any other users of the program knew if it was possible to use both single & multi line comment methods based on single line selected vs multi lines selected or even a separate hotkey assigned to single comment and multi line comment.
Again I apologize.

I had already checked the plugin section when you mentioned it but nothing came up with my keyword search "comment"

If there are any other users who can share some information about any of the above topics I am still be interested.
Thank You

**Comment by elextr** [comment]
> "sensible value is present" 

Means the `comment_single=..` setting has a value, not when it has an empty setting `comment_single=` which unsets values set in overridden system defaults.

It is important to note that you are using the `Toggle Line Commentation` which is different to the `Comment Line(s)` and `Uncomment Line(s)` since toggle adds the `Comment Toggle Marker` (default preference `~`) to the comment so toggled lines are identifiable for untoggle actions.   The comment and uncomment actions purely use the comment.

Thats why its important to clearly define what command you are using and the actual command by name, which tells readers what it does more clearly than a keybinding (eg "Toggle Line Commentation" has meaning vs Ctrl+E which has no meaning), and since the keybinding may be modified or defined by a user it is better not to use it.

There are only actions for Toggle comment, Comment line, and Uncomment line functions which have subtleties in actions depending on which comment settings are present and which needs someone to check the code what, why, where, etc.

It may be more obvious to have separate commands specifying use of single line or multiple token comment actions and the subtleties of behaviour can be removed.  PRs may be welcome.

**Comment by Skif-off** [comment]
> I found the `filetypes.lua` file ... So it is set for single / multiple lines

If I understand correctly, this file describes what symbols and words to use for parsing and highlighting code. And "Comment line(s)" & "Uncomment line(s)" functions (line by line) are just a nice addition. (+ Edit > Insert Comments > Insert Multiline Commemt.) Also these functions have a special feature, they use "~" as an indicator of their work: these are simple functions and you shouldn't expect miracles.

"Removing multiline comments" (i.e. `--[[ ]]`, `--[=[ ]=]`, `--[==[ ]==]` and so on) sounds like... hmm... a big adventure :)

P.S. "Insert Multiline Commemt" works sadly. Maybe GeanyLua can help in this case?

**Comment by advice2020** [comment]
elextr & Skif-off :)
Thank you guys for the responses.
The documentation does not really mention about the comment toggle working with the marker symbol.
I had actually removed the marker symbol in the preferences, I did not realize that is what its purpose was, I just like the toggle.
As mentioned I thought the toggle was just a combination of the other two comment actions.
Seems strange that two settings are available and one is set to override the other.
As mentioned to me it seems like if the cursor was in a line then it would use single comment and if user highlights multiple lines then the same action would provide the multi line comment for those lines.  I tried a program out yesterday and was surprised that it had this as well.
I was not even thinking of the Edit > Insert Comments > Insert Multiline Comment, this one does not have an action available in the shortcuts list.
I had some ideas that would cover all the different commenting actions but without developers to add them, probably not worth mentioning at this time.
Anyways, thanks again guys for the responses.

**Comment by elextr** [comment]
> The documentation does not really mention about the comment toggle working with the marker symbol.

True the document is only relative to the pref for the marker, not mentioned on the document on the action, feel free to provide a PR to add such linking from the action because so far nobody has needed it.

The fact that for a lot of languages the delimited comments (eg `/*` and `*/` for C/C++ and many others) __do not nest__ and as such they are a problem if the selected lines happens to include a comment.

For example if the three lines below are selected and commented by the delimiting (multiple line as you call it) comments:
```C
a = 1;
b = foo(a /*parameter 1*/, 55, "whatever");
a = a + b;
```
the selected three lines are not removed as expected, instead the code seen by the compiler is:
```C
,55, "whatever");
a = a + b; */
```
which of course does not compile.

Whereas single line comments work fine using `Comment` and `Uncomment`.

But toggle has problems too, if you have no specifier of a marker it just toggles the three selected lines:

```C
pay(user_salary, monthly, to_user);
// pay(user_salary, daily, to_me);
pay(rent);
```
to
```C
// pay(user_salary, monthly, to_user);
pay(user_salary, daily, to_me);
// pay(rent);
```
which is probably not what you intended but pays me lots ;-).

If a marker is specified (eg default ~), toggle will not toggle comments that do not have the marker, so the result is:

```C
//~ pay(user_salary, monthly, to_user);
// pay(user_salary, daily, to_me);
//~ pay(rent);
```
and toggles back to

```C
pay(user_salary, monthly, to_user);
// pay(user_salary, daily, to_me);
pay(rent);
```

**Comment by advice2020** [comment]
Thank you for your response / examples.
I understand, makes it even trickier because each language uses different methods for commenting as well as some languages using more than one symbol to comment items out.
So far the toggle hotkey was working for me but I have only been testing / languages have only supported commenting out the whole lines rather than nested.  But it does toggle lines if there are any in-between when applying to multiple lines at once.
I do not have it installed anymore, but as I mentioned I tried a program a while ago that I was surprised offered this feature.  I am curious how that program handled it, but at the time I didn't test it out that much and was not thinking about what the two of you guys have shared.
Anyways, thank you again for your response, and although it was informative, I am not paying you lots! :)

**Comment by Skif-off** [comment]
> I tried a program a while ago that I was surprised offered this feature.

Name?

**Comment by advice2020** [comment]
@Skif-off
I did this on purpose because I do not know how others feel about mentioning other program names, but since you asked specifically, I believe the programs is called “Bluefish”.
I think this is the one that allowed for both commenting methods in languages like Lua that contain two.
Just keep in mind as mentioned when I had tested commenting in this application in the past I was not aware of some of the issues that you guys shared, so not sure what the results will be if you decide to try it.
If you do try it, if it is okay with developers on this page, would be interested in hearing if it does work with both symbol/commenting methods.

Thank You
